tools/check_doc.py

```diff
@@ -4,98 +4,173 @@
-
-    os.unlink(SPEC_DIR)
+def parse_code_block_type(type_str: str) -> CodeBlockType:
+    normalized = type_str.lower()
+    types = {
+        "rflx": CodeBlockType.RFLX,
+        "ada": CodeBlockType.ADA,
+        "python": CodeBlockType.PYTHON,
+        "yaml": CodeBlockType.YAML,
+        "ignore": CodeBlockType.IGNORE,
+    }
+    if normalized not in types:
+        return CodeBlockType.UNKNOWN
+    return types[normalized]
 
-    return valid
 
+def check_file(filename: Path, content: str) -> None:
+    # pylint: disable=too-many-branches, too-many-statements
 
-def check_code(block: str, block_type: CodeBlockType, subtype: str = None) -> bool:
-    if block_type == CodeBlockType.RFLX:
-        assert subtype is not None
-        return check_rflx_code(block, subtype)
-    if block_type is CodeBlockType.ADA:
-        return check_ada_code(block)
-    if block_type is CodeBlockType.PYTHON:
-        return check_python_code(block)
-    return True
+    found = False
+    state = State.OUTSIDE
+    block = ""
+    block_start: Optional[int] = None
+    doc_type: Optional[CodeBlockType] = None
+    subtype: Optional[str] = None
+
+    for lineno, line in enumerate(content.splitlines()):
+        if state == State.INSIDE:
+            match = re.match(r"^\S", line)
+            if match:
+                check_code(filename, block_start, block, doc_type, subtype)
+                state = State.OUTSIDE
+                doc_type = None
+                subtype = None
+                block_start = None
+                block = ""
+                found = True
+            else:
+                block += f"{line}\n"
+
+            # fall-through: continue matching this line as it may already be the start of
+            # the next block
+
+        match = re.match(r"^\s*\.\. code-block::", line)
+        if match:
+            raise CheckDocError(
+                (f"{filename}:{lineno}: code-block directive forbidden (use 'code::' instead)")
+            )
+
+        match = re.match(r"^\s*\.\. doc-check: (?P<type>\S+)\s*$", line)
+        if match:
+            state = State.HEADER
+            check = match.group("type").split(",")
+            doc_type = parse_code_block_type(check[0])
+            if len(check) > 1:
+                subtype = check[1]
+
+            continue
+
+        match = re.match(r"^\s*\.\. code:: (?P<tag>\S+)\s*$", line)
+        if match:
+            block_type = parse_code_block_type(match.group("tag"))
+            if doc_type:
+                if (
+                    (doc_type != CodeBlockType.RFLX or block_type != CodeBlockType.ADA)
+                    and doc_type != CodeBlockType.IGNORE
+                    and doc_type != block_type
+                ):
+                    raise CheckDocError(
+                        f"{filename}:{lineno}: "
+                        "inconsistent code block type "
+                        f"(block: {block_type}, doc: {doc_type})"
+                    )
```

Comment:
Is it intended that unknown doc-check directives do not lead to an error? For example:

```rst
.. doc-check: foo
.. code:: foo
```

Reply:
Good catch. Invalid doc-check types are now detected (tests added).